packages/backend/ubuntu-pro.go

```diff
@@ -0,0 +1,163 @@
+package main
+import (
+    "os/exec"
+    "context"
+    "errors"
+    pb "github.com/canonical/desktop-security-center/packages/proto"
+    "google.golang.org/protobuf/types/known/emptypb"
+    "github.com/tidwall/gjson"
+    "github.com/godbus/dbus/v5"
+)
+
+var conn *dbus.Conn
+var reqId string
+
+func isServiceEnabled(basename string) (bool, error) {
+    obj := conn.Object(
+        "com.canonical.UbuntuAdvantage",
+        dbus.ObjectPath("/com/canonical/UbuntuAdvantage/Services/" + basename),
+    )
+    status, err := obj.GetProperty("com.canonical.UbuntuAdvantage.Service.Status")
+    if err != nil {
+        return false, err
+    }
+    return status.Value().(string) == "enabled", nil
+}
+
+func (s *ProServer) IsMachineProAttached(ctx context.Context, _ *emptypb.Empty) (*pb.Boolean, error) {
+    obj := conn.Object(
+        "com.canonical.UbuntuAdvantage",
+        "/com/canonical/UbuntuAdvantage/Manager",
+    )
+    isAttached, err := obj.GetProperty("com.canonical.UbuntuAdvantage.Manager.Attached")
+    if err != nil {
+        return nil, err
+    }
+
+    return &pb.Boolean{ Ret: isAttached.Value().(bool) }, nil
+}
+
+func (s *ProServer) IsEsmInfraEnabled(ctx context.Context, _ *emptypb.Empty) (*pb.Boolean, error) {
+    enabled, err := isServiceEnabled("esm_2dinfra")
+    return &pb.Boolean{ Ret: enabled }, err
+}
+
+func (s *ProServer) IsEsmAppsEnabled(ctx context.Context, _ *emptypb.Empty) (*pb.Boolean, error) {
+    enabled, err := isServiceEnabled("esm_2dapps")
+    return &pb.Boolean{ Ret: enabled }, err
+}
+
+func (s *ProServer) IsKernelLivePatchEnabled(ctx context.Context, _ *emptypb.Empty) (*pb.Boolean, error) {
+    enabled, err := isServiceEnabled("livepatch")
+    return &pb.Boolean{ Ret: enabled }, err
+}
+
+var dbusServices map[dbus.ObjectPath]map[string]map[string]dbus.Variant
+func connectToSystemBus() error {
+    var err error
+    conn, err = dbus.ConnectSystemBus()
+    return err
+}
+
+func enableService(basename string, able string) error {
+    obj := conn.Object(
+        "com.canonical.UbuntuAdvantage",
+        dbus.ObjectPath("/com/canonical/UbuntuAdvantage/Services/" + basename),
+    )
+    call := obj.Call(
+        "com.canonical.UbuntuAdvantage.Service." + able,
+        dbus.FlagAllowInteractiveAuthorization,
+    )
+    if call.Err != nil {
+        return call.Err
+    }
+    return nil
+}
+
+func (s *ProServer) EnableKernelLivePatch(ctx context.Context, _ *emptypb.Empty) (*emptypb.Empty, error) {
+    return new(emptypb.Empty), enableService("livepatch", "Enable")
```

Comment:
It generally more idiomatic to return nil as your response if an error has occurred. So I would rather suggest:

```
...
err := enableService("livepatch", "Enable") 
if err != nil {
    // Option to add logging if needed
    return nil, err
}
return new(emptypb.Empty), nil
...
```
Applied to all calls like this one

Reply:
Grpcurl always gives me this error if I try it to return nil where an Empty is expected:

```
ERROR:
  Code: Internal
  Message: grpc: error while marshaling: proto: Marshal called with nil
```

Reply:
Try using [grpc status's](https://grpc.github.io/grpc/core/md_doc_statuscodes.html), I think that will solve your issue:

```
import (
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

...

if err != nil {
	return nil, status.Errorf(codes.Internal, "failed to preform this task: %v", err)
}
```